chain/test_chain_tracer.go

```diff
@@ -1,118 +1,150 @@
-func (t *TestChainTracerRouter) CaptureEnd(output []byte, gasUsed uint64, err error) {
+// OnExit is called after a call to finalize tracing completes for the top of a call frame, as defined by tracers.Tracer.
+func (t *TestChainTracerRouter) OnExit(depth int, output []byte, gasUsed uint64, err error, reverted bool) {
 	// Call the underlying method for each registered tracer.
 	for _, tracer := range t.tracers {
-		tracer.CaptureEnd(output, gasUsed, err)
+		if tracer.OnExit != nil {
+			tracer.OnExit(depth, output, gasUsed, err, reverted)
+		}
 	}
 }
 
-// CaptureEnter is called upon entering of the call frame, as defined by vm.EVMLogger.
-func (t *TestChainTracerRouter) CaptureEnter(typ vm.OpCode, from common.Address, to common.Address, input []byte, gas uint64, value *big.Int) {
+// OnOpcode records data from an EVM state update, as defined by tracers.Tracer.
+func (t *TestChainTracerRouter) OnOpcode(pc uint64, op byte, gas, cost uint64, scope tracing.OpContext, rData []byte, depth int, err error) {
 	// Call the underlying method for each registered tracer.
 	for _, tracer := range t.tracers {
-		tracer.CaptureEnter(typ, from, to, input, gas, value)
-	}
-}
+		if tracer.OnOpcode != nil {
 
-// CaptureExit is called upon exiting of the call frame, as defined by vm.EVMLogger.
-func (t *TestChainTracerRouter) CaptureExit(output []byte, gasUsed uint64, err error) {
-	// Call the underlying method for each registered tracer.
-	for _, tracer := range t.tracers {
-		tracer.CaptureExit(output, gasUsed, err)
+			tracer.OnOpcode(pc, op, gas, cost, scope, rData, depth, err)
+		}
 	}
 }
 
-// CaptureState records data from an EVM state update, as defined by vm.EVMLogger.
-func (t *TestChainTracerRouter) CaptureState(pc uint64, op vm.OpCode, gas, cost uint64, scope *vm.ScopeContext, rData []byte, depth int, vmErr error) {
+// OnFault records an execution fault, as defined by tracers.Tracer.
+func (t *TestChainTracerRouter) OnFault(pc uint64, op byte, gas, cost uint64, scope *vm.ScopeContext, depth int, err error) {
 	// Call the underlying method for each registered tracer.
 	for _, tracer := range t.tracers {
-		tracer.CaptureState(pc, op, gas, cost, scope, rData, depth, vmErr)
+		if tracer.OnFault != nil {
+			tracer.OnFault(pc, op, gas, cost, scope, depth, err)
+		}
 	}
 }
 
-// CaptureFault records an execution fault, as defined by vm.EVMLogger.
-func (t *TestChainTracerRouter) CaptureFault(pc uint64, op vm.OpCode, gas, cost uint64, scope *vm.ScopeContext, depth int, err error) {
+func (t *TestChainTracerRouter) OnCodeChange(a common.Address, prevCodeHash common.Hash, prev []byte, codeHash common.Hash, code []byte) {
 	// Call the underlying method for each registered tracer.
+
 	for _, tracer := range t.tracers {
-		tracer.CaptureFault(pc, op, gas, cost, scope, depth, err)
+		if tracer.OnCodeChange != nil {
+			tracer.OnCodeChange(a, prevCodeHash, prev, codeHash, code)
+		}
 	}
+
 }
```

Comment:
we probably don't need to subscribe to this right now since we don't have the `OnCodeChange` hook being used anywhere?